### ISSUE TITLE
Clarification on duplicate anchor precedence §3.2.2.2

### ISSUE THREAD

**Issue by am11** [yaml/yaml-spec#48]
In the following excerpt from [`§3.2.2.2`](https://yaml.org/spec/1.2/spec.html#id2765878):

> When composing a representation graph from serialized events, an alias node refers to the most recent node in the serialization having the specified anchor. Therefore, anchors need not be unique within a serialization.

what is the correct interpretation of <kbd>the most recent node</kbd>?

Given this input:

```yaml
a: &anchor1 test0
b:
  b1: &anchor1 test1
  b2:
    b21:  &anchor1 test2
c:  *anchor1
```

the ambiguity is that in case of depth-first traversal, the value of `c` would be `test2`, otherwise `test0` in case of breadth-traversal.

**Comment by eemeli** [comment]
It's depth-first, so `*anchor1` resolves to `test2`.

**Comment by am11** [comment]
Does it mean that the spec mandates depth-first traversal somewhere?

**Comment by eemeli** [comment]
It may not say so quite explicitly, but yes, depth-first traversal is effectively required in order to produce the serialized events that section also refers to.

**Comment by perlpunk** [comment]
I agree, it's not very clear, but "most recent node in the serialization" is at least a hint that it's depth-first.
I think it also makes sense from a usage perspective. It's more likely that repeated anchors are used in independent subtrees, not inside of the same tree.

**Comment by perlpunk** [comment]
I think we don't have a test in yaml-test-suite yet that makes that behaviour clear

**Comment by am11** [comment]
I agree, a test would be helpful.

Slightly related:
I am looking at a case where alias is used before the first declaration of anchor:

```yaml
z: *anchor1
a: &anchor1 test0
b:
  b1: &anchor1 test1
  b2:
    b21:  &anchor1 test2
c: *anchor1
```

going by the spec and depth-first traversal, I think `z` should have the value `test0` while `c` is `test2`. some implementations support such formation ([ref](https://dotnetfiddle.net/ZpRHoc)), others don't ([ref](https://jsonformatter.org/yaml-to-json/988a48)). It would be helpful if a test covers this scenario as well; whether or not we should error out.

**Comment by eemeli** [comment]
Your example should throw an error, as the `*anchor1` value cannot be resolved within the z value. On this the spec is clear (Section 7.1):
> It is an error for an alias node to use an anchor that does not previously occur in the document.

**Comment by perlpunk** [comment]
One reason we don't have a test for that so far is, that the parser ideally doesn't keep track of anchors, so for the parser this is valid syntax. It should only error at a later point, when composing the data structure. But we don't have a distinction for this type of error yet.

**Comment by ingydotnet** [comment]
Sorry to be late to this party. I'm reviewing issues for the upcoming 1.2.1 spec for YAML 1.2...

TL;DR `c == test2`

Given:
```
a: &anchor1 test0
b:
  b1: &anchor1 test1
  b2:
    b21:  &anchor1 test2
c:  *anchor1
```

Every parser in https://spec.yaml.io/playground/parser parses that into these events:
```
+STR
+DOC
+MAP
=VAL :a
=VAL &anchor1 :test0
=VAL :b
+MAP
=VAL :b1
=VAL &anchor1 :test1
=VAL :b2
+MAP
=VAL :b21
=VAL &anchor1 :test2
-MAP
-MAP
=VAL :c
=ALI *anchor1
-MAP
-DOC
-STR
```

A composer keeps an internal mapping of alias names to graph nodes, replacing the value when a duplicate anchor key is received. There is no "depth" in play here, it's just a stream of events. Thus it it is obvious that the nearest `&anchor1` value is `test2`.

The spec says:
> 3.2.2.2. Anchors and Aliases  
In the representation graph, a node may appear in more than one collection. When serializing such data, the first occurrence of the node is identified by an anchor. Each subsequent occurrence is serialized as an alias node which refers back to this anchor. Otherwise, anchor names are a serialization detail and are discarded once composing is completed. **When composing a representation graph from serialized events, an alias node refers to the most recent node in the serialization having the specified anchor.** Therefore, anchors need not be unique within a serialization. In addition, an anchor need not have an alias node referring to it. It is therefore possible to provide an anchor for all nodes in serialization.

Here "serialization" means stream of events.

In my opinion, it would be clearer to say "the most recent **event** in the serialization". The term "node" is used a bit generically in the spec. It could mean a native object, a composed graph object or an event object. Since we are in the context of a serialization here (a stream of events), "node" should be disambiguated to "event".

**Comment by ingydotnet** [comment]
See #106

Currently viewable here in 1.2.1 wip spec: https://ingydotnet.github.io/yaml-spec/spec#3222-anchors-and-aliases

**Comment by ingydotnet** [comment]
Resolved in #106

**Comment by am11** [comment]
Thanks everyone for the clarification. 8-)
